media/js/mozorg/contribute/taskview.js

```diff
@@ -0,0 +1,316 @@
+     */
+    function getVisibilityStateEventKeyword() {
+        // for browser that support the event unprefixed
+        if (typeof document.hidden !== 'undefined') {
+            return 'visibilitychange'
+        }
+
+        // for Chrome 13 and lower
+        if (typeof document.webkitHidden !== 'undefined') {
+            return 'webkitvisibilitychange'
+        }
+
+        // for older versions of IE
+        if (typeof document.msHidden !== 'undefined') {
+            return 'msvisibilitychange'
+        }
+    }
+
+    /**
+     * Toggles the completed class on the relevant step, and calls
+     * taskComplete() if this was the final step of the task.
+     */
+    function completeStep($step) {
+        var $stepOne = $('#step_one');
+        var $stepTwo = $('#step_two');
+
+        if ($step.data('step') === 'one') {
+            $stepOne.toggleClass('completed');
+        } else if ($step.data('step') === 'two') {
+            $stepTwo.toggleClass('completed');
+        }
+
+        if ($step.data('complete') === true) {
+            taskComplete();
+        }
+    }
+
+    /**
+     * Called once all steps of the task has been completed. This will
+     * show the thank message and scroll it into view.
+     */
+    function taskComplete() {
+        $thankYou.removeClass('visibly-hidden');
+        $thankYou.attr('aria-hidden', 'false');
+        $thankYou[0].scrollIntoView();
+        $thankYou.focus();
+    }
+
+    /**
+     * Waits for the initial tab/window to become visible, and then
+     * proceeds to complete the relevant task step.
+     */
+    function handleVisibilityChange($step) {
+
+        var $document = $(document);
+        var visibilityChange = getVisibilityStateEventKeyword();
+
+        // to be sure we do not queue a bunch of visibilityChange events,
+        // because the browser did not fire the event, and thus `.off`
+        // was never called, we first ensue no events are currently bound,
+        // before binding a new one.
+        $document.off(visibilityChange + '.taskview');
+
+        $document.on(visibilityChange + '.taskview', function() {
+            // we wait until our current tab is visible before
+            // showing the thank you message.
+            if (document.visibilityState === 'visible') {
+                completeStep($step);
+                $document.off(visibilityChange + '.taskview');
+            }
+        });
+    }
+
+    /**
+     * Handles blur and focus events on the main window, and completes
+     * the task step once the main window receives focus.
+     */
+    function handleFocusChange($step) {
+        var taskCompleted;
+
+        if (!taskCompleted) {
+            // once our original window receives focus again, complete the task.
+            window.onfocus = function() {
+                completeStep($step);
+            };
+            // ensure this only happens once.
+            taskCompleted = true;
+        }
+    }
```

Comment:
The logic in this function still isn't quite right. As it is function scoped, `taskCompleted` will be falsy every time the function runs, which results in the `onfocus` handler being called each time the original window regains focus (e.g. when switching back and forth from the Twitter popup).

A better approach may be to use jQuery's `one` function to make sure the handler only fires once. Something like:

```
$(window).one('focus', function() {
    completeStep($step);
});
```